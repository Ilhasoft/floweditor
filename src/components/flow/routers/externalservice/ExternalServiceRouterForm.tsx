import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { nodeToState, stateToNode } from './helpers';
import { FormState, mergeForm, StringEntry, FormEntry } from 'store/nodeEditor';
import {
  Alphanumeric,
  Required,
  shouldRequireIf,
  StartIsNonNumeric,
  validate
} from 'store/validators';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import { Asset } from 'store/flowContext';
import * as React from 'react';
import i18n from 'config/i18n';
import TembaSelect from 'temba/TembaSelect';
import TypeList from 'components/nodeeditor/TypeList';
import { createResultNameInput } from 'components/flow/routers/widgets';
import ParamList from 'components/flow/routers/paramlist/ParamList';
import { ServiceCall } from 'config/interfaces';
import { ServicesCalls } from './constants';

export interface ExternalServiceRouterFormState extends FormState {
  externalService: FormEntry;
  call: FormEntry;
  resultName: StringEntry;
  calls: FormEntry;
  params: FormEntry;
}

export default class ExternalServiceRouterForm extends React.Component<
  RouterFormProps,
  ExternalServiceRouterFormState
> {
  constructor(props: RouterFormProps) {
    super(props);
    const externalServices = Object.values(this.props.assetStore.externalServices.items);
    const externalService = externalServices.length === 1 ? externalServices[0] : null;
    this.state = nodeToState(this.props.nodeSettings, externalService);

    bindCallbacks(this, {
      include: [/^handle/]
    });
  }
  private handleUpdate(
    keys: {
      externalService?: Asset;
      call?: ServiceCall;
      resultName?: string;
      calls?: any[];
      params?: any[];
    },
    submitting = false
  ): boolean {
    const updates: Partial<ExternalServiceRouterFormState> = {};

    if (keys.hasOwnProperty('externalService')) {
      updates.externalService = validate(
        i18n.t('forms.externalService', 'External Service'),
        keys.externalService,
        [shouldRequireIf(submitting)]
      );
    }

    if (keys.hasOwnProperty('call')) {
      updates.call = validate(i18n.t('forms.call', 'Call'), keys.call, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate(i18n.t('forms.result_name', 'Result Name'), keys.resultName, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('calls')) {
      updates.calls = validate(i18n.t('forms.calls', 'Calls'), keys.calls, [
        shouldRequireIf(submitting)
      ]);
    }

    // TODO: Improve paramter validation to check each field
    if (keys.hasOwnProperty('params')) {
      updates.params = validate(i18n.t('forms.params', 'Params'), keys.params, [
        shouldRequireIf(submitting)
      ]);
    }

    const updated = mergeForm(this.state, updates);

    this.setState(updated);
    return updated.valid;
  }

  private handleExternalServiceUpdate(selected: Asset[]): void {
    const newService = selected[0];
    this.handleUpdate({ externalService: newService, calls: ServicesCalls[newService.type] || [] });
  }

  private handleCallUpdate(call: ServiceCall): void {
    this.handleUpdate({ call });
  }

  private handleParamsUpdated(params: any[]): void {
    this.handleUpdate({ params });
  }

  private handleResultNameUpdate(value: string): void {
    const resultName = validate(i18n.t('forms.result_name', 'Result Name'), value, [
      Required,
      Alphanumeric,
      StartIsNonNumeric
    ]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  private handleSave(): void {
    // remove the last param as it is an empty one
    this.state.params.value.pop();

    const valid = this.handleUpdate(
      {
        externalService: this.state.externalService.value,
        call: this.state.call.value,
        params: this.state.params.value,
        resultName: this.state.resultName.value
      },
      true
    );

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  private renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const showExternalServices =
      Object.keys(this.props.assetStore.externalServices.items).length > 0 ||
      this.props.issues.length > 0;

    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        {showExternalServices ? (
          <div>
            <p>
              <span>Integration with... </span>
            </p>
            <AssetSelector
              key="select_external_service"
              name={i18n.t('forms.external_service', 'External Service')}
              placeholder="Select the external service to use"
              assets={this.props.assetStore.externalServices}
              onChange={this.handleExternalServiceUpdate}
              entry={this.state.externalService}
            />
          </div>
        ) : (
          ''
        )}
        <div>
          <p>
            <span>Action</span>
          </p>
          <TembaSelect
            key="select_external_service_call"
            name={i18n.t('forms.external_service_call', 'External Service Call')}
            placeholder="Select the call from external service to do"
            options={this.state.calls.value}
            nameKey="verboseName"
            onChange={this.handleCallUpdate}
            value={this.state.call.value}
            searchable={false}
          />
        </div>
        {this.state.call.value ? (
          <ParamList
            availableParams={this.state.call.value.params}
            params={this.state.params.value}
            onParamsUpdated={this.handleParamsUpdated}
          />
        ) : (
          <></>
        )}
        {createResultNameInput(this.state.resultName, this.handleResultNameUpdate)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
