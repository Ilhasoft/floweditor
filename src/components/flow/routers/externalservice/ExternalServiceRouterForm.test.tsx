import { createCallExternalServiceNode, getRouterFormProps } from 'testUtils/assetCreators';
import { Types } from 'config/interfaces';
import { RenderNode } from 'store/flowContext';
import ExternalServiceRouterForm from './ExternalServiceRouterForm';
import * as React from 'react';
import {
  render,
  fireEvent,
  fireChangeText,
  fireTembaSelect,
  fireUnnnicInputChangeText,
  fireUnnnicSelect
} from 'test/utils';
import { act } from '@testing-library/react';

const externalServiceAsset = require('test/assets/external_services.json');

// eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
const externalServiceForm = getRouterFormProps({
  node: createCallExternalServiceNode('omie'),
  ui: { type: Types.split_by_external_service }
} as RenderNode);

describe(ExternalServiceRouterForm.name, () => {
  describe('render', () => {
    it('should render', async () => {
      let rendered: any;
      await act(async () => {
        rendered = render(<ExternalServiceRouterForm {...externalServiceForm} />);
        return undefined;
      });
      expect(rendered.baseElement).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('should save changes', async () => {
      externalServiceForm.assetStore.externalServices.items = {
        [externalServiceAsset.results[0].uuid]: externalServiceAsset.results[0],
        [externalServiceAsset.results[1].uuid]: externalServiceAsset.results[1]
      };
      let rendered: any;

      await act(async () => {
        rendered = render(<ExternalServiceRouterForm {...externalServiceForm} />);
        return undefined;
      });

      expect(rendered.baseElement).toMatchSnapshot();

      const okButton = rendered.getByText('Ok');
      const resultName = rendered.getByTestId('Save as result');

      fireUnnnicInputChangeText(resultName, '');
      fireEvent.click(okButton);
      expect(externalServiceForm.updateRouter).not.toBeCalled();

      await act(async () => {
        fireUnnnicSelect(
          rendered.getByTestId('temba_select_external_service'),
          { value: [externalServiceAsset.results[1]] },
          'value'
        );
      });

      await act(async () => {
        fireUnnnicSelect(
          rendered.getByTestId('temba_select_external_service_call'),
          {
            value: [
              {
                name: 'IncluirContato',
                value: 'IncluirContato',
                verboseName: 'Inserir Contato'
              }
            ]
          },
          'value'
        );
      });

      await act(async () => {
        fireUnnnicInputChangeText(resultName, 'My External Service Result');
      });

      await act(async () => {
        fireUnnnicInputChangeText(rendered.getAllByTestId('Service Call Param Data')[0], 'data 1');
      });

      // cannot be called due to missing second required field
      expect(externalServiceForm.updateRouter).not.toBeCalled();

      await act(async () => {
        fireUnnnicInputChangeText(rendered.getAllByTestId('Service Call Param Data')[1], 'data 2');
      });

      fireEvent.click(okButton);
      expect(externalServiceForm.updateRouter).toHaveBeenCalled();
      expect(externalServiceForm.updateRouter).toMatchCallSnapshot();
    });
  });
});
