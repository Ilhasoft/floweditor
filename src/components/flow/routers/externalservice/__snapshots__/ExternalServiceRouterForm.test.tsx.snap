// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalServiceRouterForm render should render 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header call_external_service"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Call External Service
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"call_external_service\\",\\"name\\":\\"Call External Service\\",\\"description\\":\\"Call a external service\\",\\"localizeableKeys\\":[\\"exists\\"],\\"aliases\\":[\\"split_by_external_service\\"],\\"filter\\":\\"external_service\\"}]"
              />
            </div>
          </div>
        </div>
        
        <div>
          <p>
            <span>
              Action
            </span>
          </p>
          <div
            class="normal "
          >
            <temba-select
              data-testid="temba_select_external_service_call"
              errors="[]"
              expressions=""
              name="External Service Call"
              namekey="verboseName"
              placeholder="Select the call from external service to do"
              valuekey="value"
              values="[]"
            />
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            data-use-vue-component-wrap=""
            style="all: unset;"
          >
            <div
              class="unnnic-text-area sm normal"
              data-passed-props="[object Object]"
              data-v-a95e6368=""
            >
              <div
                class="label"
                data-v-a95e6368=""
              >
                Result Name
              </div>
              <textarea
                data-v-a95e6368=""
                maxlength="64"
              />
              <div
                class="helper"
                data-v-a95e6368=""
              >
                 6/64 
              </div>
              <!---->
            </div>
          </div>
          By naming the result, you can reference it later using @results.result
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ExternalServiceRouterForm updates should save changes 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header call_external_service"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Call External Service
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"call_external_service\\",\\"name\\":\\"Call External Service\\",\\"description\\":\\"Call a external service\\",\\"localizeableKeys\\":[\\"exists\\"],\\"aliases\\":[\\"split_by_external_service\\"],\\"filter\\":\\"external_service\\"}]"
              />
            </div>
          </div>
        </div>
        <div>
          <p>
            <span>
              Integration with... 
            </span>
          </p>
          <div
            class="ele"
          >
            <div
              class="normal "
            >
              <temba-select
                cachekey="0"
                data-testid="temba_select_external_service"
                errors="[]"
                expressions=""
                name="External Service"
                namekey="name"
                placeholder="Select the external service to use"
                valuekey="uuid"
                values="[{\\"uuid\\":\\"4b154b06-5ecd-43d9-afca-39738e6859d7\\",\\"name\\":\\"omie dummy project\\",\\"external_service_type\\":\\"omie\\"}]"
              />
            </div>
          </div>
        </div>
        <div>
          <p>
            <span>
              Action
            </span>
          </p>
          <div
            class="normal "
          >
            <temba-select
              data-testid="temba_select_external_service_call"
              errors="[]"
              expressions=""
              name="External Service Call"
              namekey="verboseName"
              placeholder="Select the call from external service to do"
              valuekey="value"
              values="[]"
            />
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            data-use-vue-component-wrap=""
            style="all: unset;"
          >
            <div
              class="unnnic-text-area sm normal"
              data-passed-props="[object Object]"
              data-v-a95e6368=""
            >
              <div
                class="label"
                data-v-a95e6368=""
              >
                Result Name
              </div>
              <textarea
                data-v-a95e6368=""
                maxlength="64"
              />
              <div
                class="helper"
                data-v-a95e6368=""
              >
                 6/64 
              </div>
              <!---->
            </div>
          </div>
          By naming the result, you can reference it later using @results.result
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ExternalServiceRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "call": Object {
            "name": "IncluirContato",
            "params": Array [
              Object {
                "filters": Array [
                  Object {
                    "name": "nCod",
                    "type": "integer",
                    "verboseName": "Código do Contato",
                  },
                  Object {
                    "name": "cCodInt",
                    "required": true,
                    "type": "text",
                    "verboseName": "Código de Integração",
                  },
                  Object {
                    "name": "cNome",
                    "required": true,
                    "type": "string",
                    "verboseName": "Nome do contato",
                  },
                  Object {
                    "name": "cSobrenome",
                    "type": "string",
                    "verboseName": "Sobrenome do contato",
                  },
                  Object {
                    "name": "cCargo",
                    "type": "string",
                    "verboseName": "Cargo",
                  },
                  Object {
                    "name": "dDtNasc",
                    "type": "string",
                    "verboseName": "Data de Nascimento",
                  },
                  Object {
                    "name": "nCodVend",
                    "type": "integer",
                    "verboseName": "Código do Vendedor",
                  },
                  Object {
                    "name": "nCodConta",
                    "type": "integer",
                    "verboseName": "Código da Conta",
                  },
                ],
                "type": "identificacao",
                "verboseName": "Identificação",
              },
              Object {
                "filters": Array [
                  Object {
                    "maxLength": 60,
                    "name": "cEndereco",
                    "type": "string",
                    "verboseName": "Endereço do contato",
                  },
                  Object {
                    "maxLength": 200,
                    "name": "cCompl",
                    "type": "string",
                    "verboseName": "Complemento ",
                  },
                  Object {
                    "maxLength": 10,
                    "name": "cCEP",
                    "type": "string",
                    "verboseName": "CEP",
                  },
                  Object {
                    "maxLength": 60,
                    "name": "cBairro",
                    "type": "string",
                    "verboseName": "Bairro ",
                  },
                  Object {
                    "maxLength": 50,
                    "name": "cCidade",
                    "type": "string",
                    "verboseName": "Cidade ",
                  },
                  Object {
                    "maxLength": 50,
                    "name": "cUF",
                    "type": "string",
                    "verboseName": "Estado ",
                  },
                  Object {
                    "maxLength": 50,
                    "name": "cPais",
                    "type": "string",
                    "verboseName": "País",
                  },
                ],
                "type": "endereco",
                "verboseName": "Endereço",
              },
              Object {
                "filters": Array [
                  Object {
                    "maxLength": 5,
                    "name": "cDDDCel1",
                    "type": "string",
                    "verboseName": "DDD do celular 1",
                  },
                  Object {
                    "maxLength": 15,
                    "name": "cNumCel1",
                    "type": "string",
                    "verboseName": "Número do Celular 1",
                  },
                  Object {
                    "maxLength": 5,
                    "name": "cDDDCel2",
                    "type": "string",
                    "verboseName": "DDD do Celular 2",
                  },
                  Object {
                    "maxLength": 15,
                    "name": "cNumCel2",
                    "type": "string",
                    "verboseName": "Número do Celular 2",
                  },
                  Object {
                    "maxLength": 5,
                    "name": "cDDDTel",
                    "type": "string",
                    "verboseName": "DDD do Telefone",
                  },
                  Object {
                    "maxLength": 15,
                    "name": "cNumTel",
                    "type": "string",
                    "verboseName": "Número do Telefone",
                  },
                  Object {
                    "maxLength": 5,
                    "name": "cDDDFax",
                    "type": "string",
                    "verboseName": "DDD do Fax",
                  },
                  Object {
                    "maxLength": 15,
                    "name": "cNumFax",
                    "type": "string",
                    "verboseName": "Número do Fax",
                  },
                  Object {
                    "maxLength": 200,
                    "name": "cEmail",
                    "type": "string",
                    "verboseName": "E-mail do contato",
                  },
                  Object {
                    "maxLength": 100,
                    "name": "cWebsite",
                    "type": "string",
                    "verboseName": "WebSite",
                  },
                ],
                "type": "telefone_email",
                "verboseName": "Telefone e Email",
              },
              Object {
                "filters": Array [],
                "type": "cObs",
                "verboseName": "Observações",
              },
            ],
            "value": "IncluirContato",
            "verboseName": "Inserir Contato",
          },
          "external_service": Object {
            "external_service_type": "omie",
            "name": "Omie dummy project 2",
            "uuid": "5b154b06-5ecd-43d9-afca-39738e6859d7",
          },
          "params": Array [
            Object {
              "data": Object {
                "validationFailures": Array [],
                "value": "data 1",
              },
              "filter": Object {
                "value": Object {
                  "name": "cCodInt",
                  "required": true,
                  "type": "text",
                  "verboseName": "Código de Integração",
                },
              },
              "filters": Array [
                Object {
                  "name": "nCod",
                  "type": "integer",
                  "verboseName": "Código do Contato",
                },
                Object {
                  "name": "cCodInt",
                  "required": true,
                  "type": "text",
                  "verboseName": "Código de Integração",
                },
                Object {
                  "name": "cNome",
                  "required": true,
                  "type": "string",
                  "verboseName": "Nome do contato",
                },
                Object {
                  "name": "cSobrenome",
                  "type": "string",
                  "verboseName": "Sobrenome do contato",
                },
                Object {
                  "name": "cCargo",
                  "type": "string",
                  "verboseName": "Cargo",
                },
                Object {
                  "name": "dDtNasc",
                  "type": "string",
                  "verboseName": "Data de Nascimento",
                },
                Object {
                  "name": "nCodVend",
                  "type": "integer",
                  "verboseName": "Código do Vendedor",
                },
                Object {
                  "name": "nCodConta",
                  "type": "integer",
                  "verboseName": "Código da Conta",
                },
              ],
              "required": undefined,
              "type": "identificacao",
              "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
              "valid": true,
              "verboseName": "Identificação",
            },
            Object {
              "data": Object {
                "validationFailures": Array [],
                "value": "data 2",
              },
              "filter": Object {
                "value": Object {
                  "name": "cNome",
                  "required": true,
                  "type": "string",
                  "verboseName": "Nome do contato",
                },
              },
              "filters": Array [
                Object {
                  "name": "nCod",
                  "type": "integer",
                  "verboseName": "Código do Contato",
                },
                Object {
                  "name": "cCodInt",
                  "required": true,
                  "type": "text",
                  "verboseName": "Código de Integração",
                },
                Object {
                  "name": "cNome",
                  "required": true,
                  "type": "string",
                  "verboseName": "Nome do contato",
                },
                Object {
                  "name": "cSobrenome",
                  "type": "string",
                  "verboseName": "Sobrenome do contato",
                },
                Object {
                  "name": "cCargo",
                  "type": "string",
                  "verboseName": "Cargo",
                },
                Object {
                  "name": "dDtNasc",
                  "type": "string",
                  "verboseName": "Data de Nascimento",
                },
                Object {
                  "name": "nCodVend",
                  "type": "integer",
                  "verboseName": "Código do Vendedor",
                },
                Object {
                  "name": "nCodConta",
                  "type": "integer",
                  "verboseName": "Código da Conta",
                },
              ],
              "required": undefined,
              "type": "identificacao",
              "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
              "valid": true,
              "verboseName": "Identificação",
            },
          ],
          "result_name": "My External Service Result",
          "type": "call_external_service",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "Success",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_category",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Success",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Failure",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
        ],
        "default_category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@results.my_external_service_result",
        "type": "switch",
      },
      "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_external_service",
    },
  },
]
`;
